### PR TITLE
Camera handling module

### PR DESCRIPTION
**Pull request by jedrzejpolaczek**
TODO: re-think configuration loading solution.

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  "camera": {
+    "camera_resolution": {
+      "weight": 1024,
+      "height": 768
+    },
+    "image_format": ".jpg"
+  },
+  "distance_sensor": {
+  }
+}
```
Comment (GrivIN) [inline config.json:16]: I see two config.json files without any descriptions, configuring all elements lets move to separate Pas (after docker-compose is ready)

File: src/modules/camera/src/run.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+
+async def main() -> None:
+    """
+    Waiting for message from distance sensor. On message suitable value take snap from camera and broadcast it.
+    """
+    # Connections and sockets preparation
+    # Create socket from where we will be consuming signals
+    instance_context = Context.instance()
+    consume_socket = instance_context.socket(zmq.SUB)
```
Comment (GrivIN) [inline src/modules/camera/src/run.py:17]: zmq contexts should be in separate class, initiated on start and only push/pull method should be interfaced for the main

File: src/modules/camera/src/camera_snap.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    weight = config_data["camera"]["functionality"]["resolution"]["weight"]
+    height = config_data["camera"]["functionality"]["resolution"]["height"]
+    image_format = config_data["camera"]["functionality"]["image_format"]
+
+    with picamera.PiCamera() as camera:
+        camera.resolution = (weight, height)
+        camera.start_preview()
+
+        # Camera warm-up time
+        time.sleep(2)
```
Comment (GrivIN) [inline src/modules/camera/src/camera_snap.py:21]: camera warmup should be done in init method when we initiate its own class, only capture method should be available outside (no need to wait 2 sec on all requests for an image

File: src/modules/camera/src/camera_snap.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+        # Prepare file name
+        now = datetime.now()
+        now_str = now.strftime("%d%m%Y%H%M%S")
+        snap_name = now_str + image_format
```
Comment (GrivIN) [inline src/modules/camera/src/camera_snap.py:26]: it is possible than 2 images end up at same file (two images made in one second), uuid would be better

File: src/modules/camera/src/run.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    event_bus_server_ip = config_data["event_bus"]["event_bus_server_ip"]
+    receiving_port = config_data["event_bus"]["receiving_port"]
+    sending_port = config_data["event_bus"]["sending_port"]
+
+    # Load event bus module configuration
+    topic_distance = config_data["distance_sensor"]["event_bus"]["topic"]
+    topic_camera = config_data["camera"]["event_bus"]["topic"]
+
+    # Load module logic
+    border_distance = config_data["camera"]["module_logic"]["border_distance"]
```
Comment (GrivIN) [inline src/modules/camera/src/run.py:86]: this is too complicated for our purposes, lets simply get all required options from command line or environment vars

**Review by GrivIN — DISMISSED** [review]
json config file at this moment is making our integration in docker harder, please remove it

File: src/modules/camera/src/camera_snap.py (hunk 4 above)
Comment (jedrzejpolaczek) [reply src/modules/camera/src/camera_snap.py:26]: No, the camera is just too slow.

File: src/modules/camera/src/camera_snap.py (hunk 3 above)
Comment (jedrzejpolaczek) [reply src/modules/camera/src/camera_snap.py:21]: Sounds fair.

File: src/config.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
+{
```
Comment (KonradNojman) [inline src/config.json:1]: Is this file needed if we have docker?

File: src/utils/send_event.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import argparse
```
Comment (KonradNojman) [inline src/utils/send_event.py:1]: Could you expand on what is the purpose of this file and where it is used? Maybe I missed it, but I can't find a module that imports it.

**Review by KonradNojman — APPROVED** [review]
Great job! Nice to have a Docker image and camera module set 🚀 
Code for the camera module is clear and easy to read 👍 

I left just a few organizational comments, but overall looks good to me :)

File: src/camera/code/camera.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+SOMEONE_VISIBLE = 1
+IMAGE_CAPTURED = 2
+PIC_DIRECTORY = '/tmp/'
+
+
+logger = logging.getLogger(__name__)
+
+
+if CAMERA_IMPLEMENTATION == 'cv2':
```
Comment (jedrzejpolaczek) [inline src/camera/code/camera.py:27]: Due to problems with setting up OpenCV and the simplicity of action, we need to take (taking snap) I am recommending not using this library in the current solution.

File: src/camera/code/camera.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+            resized_frame = cv2.resize(frame, (self.width, self.height), )
+            cv2.imwrite(filename=filename, img=resized_frame)
+
+        def close(self):
+            if self.camera:
+                self.camera.release()
+
+
+if CAMERA_IMPLEMENTATION == 'picamera':
+    class CameraInterface:
```
Comment (jedrzejpolaczek) [inline src/camera/code/camera.py:47]: Nice concept!

File: src/camera/code/camera.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+            # Pack topic and file name to send
+            data_packer.pack_uint(topic_camera)
+            data_packer.pack_string(snap_name.encode('utf-8'))
+
+            logger.info(f'broadcasting topic: {topic_camera} with value: {snap_name}')
+
+            # Send data
+            produce_socket.send(data_packer.get_buffer())
+            data_packer.reset()
+
```
Comment (jedrzejpolaczek) [inline src/camera/code/camera.py:147]: Linter problem: please add one new line.

**Review by jedrzejpolaczek — COMMENTED** [review]
`Request change` (i can not set this state as an author of PR):
It is a very nice solution with many elegant coding schemats (like a class for different implementation of the camera, nice one!).

The main reason for the change request is my concerns about using the `OpenCV` library. It is a very heavy and problem-making library to use when we want to only take a snap from the camera. In my opinion, `picamera` library is more than enough for this task.

File: src/camera/examples/cv2.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
+# TODO: resolve problems with installing OpenCV on raspberry pi
```
Comment (jedrzejpolaczek) [inline src/camera/examples/cv2.py:1]: I think we can drop this file.

File: src/utils/send_event.py (hunk 7 above)
Comment (GrivIN) [reply src/utils/send_event.py:1]: it is an util script to be used during development, we can send any event to event bus with is directly form terminal, it is not used by any internal mirror module directly
